p2p/conn/connection_test.go

```diff
@@ -1040,3 +1047,177 @@ func tcpNetPipe() (net.Conn, net.Conn) {
+	totalDuration     time.Duration // total duration for which messages are sent
+	sendQueueCapacity int           // send queue capacity i.e., the number of messages that can be buffered
+	sendRate          int64         // send rate in bytes per second
+	recRate           int64         // receive rate in bytes per second
+	chID              byte          // channel ID
+}
+
+func runBenchmarkTest(b *testing.B, tt testCase) {
+	b.Run(tt.name, func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			// set up two networked connections
+			// server, client := NetPipe() // can alternatively use this and comment out the line below
+			server, client := tcpNetPipe()
+			defer server.Close()
+			defer client.Close()
+
+			// prepare callback to receive messages
+			allReceived := make(chan bool)
+			receivedLoad := 0 // number of messages received
+			onReceive := func(chID byte, msgBytes []byte) {
+				receivedLoad++
+				if receivedLoad >= tt.totalMsg && tt.totalMsg > 0 {
+					allReceived <- true
+				}
+			}
+
+			cnfg := DefaultMConnConfig()
+			cnfg.SendRate = tt.sendRate
+			cnfg.RecvRate = tt.recRate
+			chDescs := []*ChannelDescriptor{{ID: tt.chID, Priority: 1,
+				SendQueueCapacity: tt.sendQueueCapacity}}
+			clientMconn := NewMConnectionWithConfig(client, chDescs,
+				func(chID byte, msgBytes []byte) {},
+				func(r interface{}) {},
+				cnfg)
+			serverChDescs := []*ChannelDescriptor{{ID: tt.chID, Priority: 1,
+				SendQueueCapacity: tt.sendQueueCapacity}}
+			serverMconn := NewMConnectionWithConfig(server, serverChDescs,
+				onReceive,
+				func(r interface{}) {},
+				cnfg)
+			clientMconn.SetLogger(log.TestingLogger())
+			serverMconn.SetLogger(log.TestingLogger())
+
+			err := clientMconn.Start()
+			require.Nil(b, err)
+			defer func() {
+				_ = clientMconn.Stop()
+			}()
+			err = serverMconn.Start()
+			require.Nil(b, err)
+			defer func() {
+				_ = serverMconn.Stop()
+			}()
+
+			// start measuring the time from here to exclude the time
+			// taken to set up the connections
+			b.StartTimer()
+			// start generating messages, it is a blocking call
+			generateAndSendMessages(clientMconn,
+				tt.messagingRate,
+				tt.totalDuration,
+				tt.totalMsg,
+				tt.msgSize,
+				tt.msg,
+				tt.chID)
+
+			// wait for all messages to be received
+			<-allReceived
+			b.StopTimer()
+		}
+	})
+}
+
+func BenchmarkMConnection_ScalingPayloadSizes_HighSendRate(b *testing.B) {
+	// One aspect that could impact the performance of MConnection and the
+	// transmission rate is the size of the messages sent over the network,
+	// especially when they exceed the MConnection.MaxPacketMsgPayloadSize (
+	// messages are sent in packets of maximum size MConnection.
+	// MaxPacketMsgPayloadSize).
+	// The test cases in this benchmark involve sending messages with sizes
+	// ranging exponentially from 1KB to 4096KB (
+	// the max value of 4096KB is inspired by the largest possible PFB in a
+	// Celestia block with 128*18 number of 512-byte shares)
```

Comment:
[question] why 18? Perhaps typo?

```suggestion
	// Celestia block with 128*128 number of 512-byte shares)
```

Reply:
Done https://github.com/celestiaorg/celestia-core/pull/1179/commits/91265109e6225e34a9f2955d6f52c12c9c2b4f0e.